README
========================================================
 MATPOWER - A MATLAB(R) Power System Simulation Package
========================================================

Version:    6.0

Website:    http://www.pserc.cornell.edu/matpower/
GitHub:     https://github.com/MATPOWER/matpower

Authors:    Ray Zimmerman               <rz10@cornell.edu>
            Carlos E. Murillo-Sanchez   <carlos_murillo@ieee.org>
            and others, see AUTHORS file

            Fri, Dec 16, 2016

Copyright (c) 1997-2016, Power Systems Engineering Research Center (PSERC)
See http://www.pserc.cornell.edu/matpower/ for more info.

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved. This file is offered as-is,
without any warranty.


--------------
 INTRODUCTION
--------------

MATPOWER is a package of MATLAB(R) M-files for solving power flow and
optimal power flow problems. It is intended as a simulation tool for
researchers and educators that is easy to use and modify. MATPOWER
is designed to give the best performance possible while keeping the code
simple to understand and modify. It was initially developed as part
of the PowerWeb project <http://www.pserc.cornell.edu/powerweb/>.

MATPOWER can be downloaded from the MATPOWER website, or cloned
from the MATPOWER GitHub project, both listed above.


--------------
 TERMS OF USE
--------------

Please see the LICENSE file for the details. But here is the summary:

- Beginning with version 5.1, the code in MATPOWER is distributed under
  the 3-clause BSD license.

- MATPOWER is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY.

- While not required by the terms of the license, we do request that
  publications derived from the use of MATPOWER explicitly acknowledge
  that fact by citing [2].

[2]  R. D. Zimmerman, C. E. Murillo-Sanchez, and R. J. Thomas,
     "MATPOWER: Steady-State Operations, Planning and Analysis Tools
     for Power Systems Research and Education," Power Systems, IEEE
     Transactions on, vol. 26, no. 1, pp. 12–19, Feb. 2011.

  Additionally, we request that publications derived from the use of
  the MATPOWER Optimal Scheduling Tool (MOST), explicitly acknowledge
  that fact by citing [4] as well as [2].

[4]  C. E. Murillo-Sanchez, R. D. Zimmerman, C. L. Anderson, and
     R. J. Thomas, "Secure Planning and Operations of Systems with
     Stochastic Sources, Energy Storage and Active Demand," Smart Grid,
     IEEE Transactions on, vol. 4, no. 4, pp. 2220–2229, Dec. 2013.
        http://dx.doi.org/10.1109/TSG.2013.2281001

Note:  Versions 4.0 through 5.0 were licensed under the GPL and versions
       prior to MATPOWER 4 use a different license.


-----------------
 GETTING STARTED
-----------------

System Requirements
-------------------
    - MATLAB(R) version 7 (R14) or later
      (available from The MathWorks, Inc. http://www.mathworks.com/), or
    - GNU Octave version 3.4 or later (free software available from
      http://www.gnu.org/software/octave/).

Installation
------------
1.  Follow the download instructions on the MATPOWER website. You
    should end up with a file named matpowerXXX.zip, where XXX depends
    on the version of MATPOWER.

2.  Unzip the downloaded file. Move the resulting matpowerXXX directory
    to the location of your choice. These files should not need to be
    modified, so it is recommended that they be kept separate from your
    own code. Let <MATPOWER> denote the path to this directory.

3.  Add the following directories to your MATLAB path:
      <MATPOWER>        - core MATPOWER functions
      <MATPOWER>/t      - test scripts for MATPOWER
      <MATPOWER>/most   - core MOST functions
      <MATPOWER>/most/t - test scripts for MOST
      (optional) subdirectories of <MATPOWER>/extras -
            additional functionality and contributed code

4.  At the MATLAB prompt, type 'test_matpower' (without the quotes) to
    run the test suite and verify that MATPOWER is properly installed
    and functioning.

Running MATPOWER
----------------
To run a simple Newton power flow on the 9-bus system specified in the
file case9.m, with the default algorithm options, at the MATLAB prompt,
type:

    runpf('case9')

To load the 30-bus system data from case30.m, increase its real power
demand at bus 2 to 30 MW, then run an AC optimal power flow with
default options, type:

    define_constants;
    mpc = loadcase('case30');
    mpc.bus(2, PD) = 30;
    runopf(mpc);

By default, the results of the simulation are pretty-printed to the
screen, but the solution can also be optionally returned in a 'results'
struct. The following example shows how simple it is, after running a DC
OPF on the 118-bus system in case118.m, to access the final objective
function value, the real power output of generator 6 and the power flow
in branch 51.

    results = rundcopf('case118');
    final_objective = results.f;
    gen6_output     = results.gen(6, PG);
    branch51_flow   = results.branch(51, PF);

For additional info, see the User's Manual and the on-line help
documentation for the various MATPOWER functions. For example:
    help runpf
    help runopf
    help mpoption
    help caseformat


---------------------------
 WHAT'S NEW IN VERSION 6.0
---------------------------

Below is a summary of the changes since version 5.1 of MATPOWER. See the
CHANGES file in the docs directory for all the gory details.

* New Open Development Model
  - MATPOWER development has moved to GitHub! The code repository is
    now publicly available to clone and submit pull requests.
    <https://github.com/MATPOWER/matpower>
  - Public issue tracker for reporting bugs, submitting patches, etc.
    <https://github.com/MATPOWER/matpower/issues>
  - Separate repositories for MOST, MIPS, MP-Test, all available
    from <https://github.com/MATPOWER/>.
  - New developer e-mail list (MATPOWER-DEV-L) to facilitate
    communication between those collaborating on MATPOWER development.
    Sign up at:
    <http://www.pserc.cornell.edu/matpower/mailinglists.html#devlist>.

* New Case Files:
  - Added 9 new case files, 8 cases ranging from 1888 to 6515 buses
    representing the French system, and a 13,659-bus case representing
    parts of the of the European high voltage transmission network,
    stemming from the Pan European Grid Advanced Simulation and State
    Estimation (PEGASE) project. Thanks again to Cedric Josz and
    colleagues from the French Transmission System Operator.
  - Added case145.m, IEEE 145 bus, 50 generator dynamic test case from
    http://www.ee.washington.edu/research/pstca/dyn50/pg_tcadd50.htm.
  - Added case33bw.m, a 33-bus radial distribution system from Baran
    and Wu.

* New Features:
  - MATPOWER Optimal Scheduling Tool (MOST) is a major new feature,
    implementing a full range of optimal power scheduling problems, from a
    simple as a deterministic, single period economic dispatch problem
    with no transmission constraints to as complex as a stochastic,
    security-constrained, combined unit-commitment and multiperiod OPF
    problem with locational contingency and load-following reserves,
    ramping costs and constraints, deferrable demands, lossy storage
    resources and uncertain renewable generation.
    See docs/MOST-manual.pdf for details.
  - General mechanism for applying modifications to an existing MATPOWER
    case. See apply_changes() and idx_ct().
  - Redesigned CPF callback mechanism to handle CPF events such as
    generator limits, nose point detection, etc. Included event log
    in CPF results.
  - Added options 'cpf.enforce_p_lims' and 'cpf.enforce_q_lims' to
    enforce generator active and reactive power limts in the
    continuation power flow.
  - Added OPF option 'opf.use_vg' to provide a convenient way to have
    the OPF respect the generator voltage setpoints specified in the
    gen matrix.
  - Experimental foundation for handling of ZIP load models in power flow
    (Newton, fast-decoupled only), continuation power flow, and optimal
    power flow (MIPS, fmincon, Knitro, IPOPT solvers only). Currently,
    ZIP loads can only be specified on a system-wide basis using the
    experimental options 'exp.sys_wide_zip_loads.pw' and
    'exp.sys_wide_zip_loads.qw'.
 - Support for quadprog() under GNU Octave.
 - New contributed extras:
    - Plot electrically meaningful drawings of a MATPOWER case using
      plot_mpc() in extras/misc, contributed by Paul Cuffe.
    - Find the maximum loadability limit of a system via an optimal power
      flow and dispatchable loads, using maxloadlim() in extras/maxloadlim,
      contributed by Camille Hamon.
    - Create a quadratically-constrained quadratic programming (QCQP)
      representation of the AC optimal power flow problem using using
      qcqp_opf() in extras/misc, contributed by Cedric Josz and colleagues.
  - New functions:
    - apply_changes() and idx_ct() provide a general mechanism for
      applying modifications to an existing MATPOWER case.
    - feval_w_path() evaluates a function located at a specified path,
      outside of the Matlab path.
    - mpopt2qpopt() provides a common interface for creating options
      struct for mi/qps_matpower() from a MATPOWER options struct.
  - New function options:
    - Option to call makeB(), makeBdc(), makePTDF(), scale_load(), and
      total_load() with full case struct (mpc) instead of individual data
      matrices (bus, branch, etc.).
    - total_load(), which now computes voltage-dependent load values,
      accepts the values 'bus' and 'area' as valid values for 'load_zone'
      argument.

* Other Improvements:
  - Changed default solver order for LP, QP, MILP, MIQP problems to move
    Gurobi before CPLEX and BPMPD after OT and GLPK.
  - Added some caching to mpoption() and made minor changes to
    nested_struct_copy() to greatly decrease the overhead added by
    mpoption() when running many small problems.
  - Added option 'cpf.adapt_step_damping' to control oscillations in
    adaptive step size control for continuation power flow.
  - Added CPF user options for setting tolerances for target lambda
    detection and nose point detection, 'cpf.target_lam_tol' and
    'cpf.nose_tol', respectively.
  - Added support for Matlab Optimization Toolbox 7.5 (R2016b).
  - Added support for MOSEK v8.x.
  - Added tests for power flow with 'pf.enforce_q_lims' option.
  - Updated network reduction code to handle cases with radially
    connected external buses.
  - Updated versions of qcqp_opf() and qcqp_opf() in extras/misc, from
    Cedric Josz.
  - Added "Release History" section to Appendix of manual.
  - Many new tests.

* Bugs fixed:
  - Fixed bug in toggle_dclines() that resulted in fatal error when used
    with OPF with reactive power costs. Thanks to Irina Boiarchuk.
  - Fixed fatal bug in update_mupq() affecting cases where QMIN is greater
    than or equal to QC1MIN and QC2MIN (or QMAX is less than or equal to
    QC1MAX and QC2MAX) for all generators. Thanks Jose Miguel.
  - Copying a field containing a struct to a non-struct field with
    nested_struct_copy() now overwrites rather than causing a fatal error.
  - Fixed a bug in psse_convert_xfmr() where conversion of data for
    transformers with CZ=3 was done incorrectly. Thanks to Jose Marin
    and Yujia Zhu.
  - Fixed a fatal bug in psse_convert_xfmr() affecting transformers with
    CW and/or CZ equal to 1. Thanks to Matthias Resch.
  - Fixed a crash in have_fcn() caused by changes in OPTI Toolbox v2.15
    (or possibly v2.12)
  - Commented out isolated bus 10287 in case3375wp.m.
  - Added code to DC OPF to return success = 0 for cases where the matrix
    is singular (e.g. islanded system without slack).
  - Fixed problem in have_fcn() where SeDuMi was turning off and leaving
    off all warnings.
  - Fixed shadow prices on variable bounds for AC OPF for fmincon,
    IPOPT, and Knitro.
  - In savecase() single quotes are now escaped properly in bus names.
  - Generator capability curve parameters that define a zero-reactive
    power line no longer cause a fatal error.
  - Bad bus numbers no longer cause a fatal error (after reporting the
    bad bus numbers) in case_info().

* Incompatible Changes:
  - Removed fairmax() from the public interface by moving it inside uopf(),
    the only place it was used.
  - Removed 'cpf.user_callback_args' option and modified
    'cpf.user_callback'.
  - Changed name of 'cpf.error_tol' option to 'cpf.adapt_step_tol'.


---------------
 DOCUMENTATION
---------------

There are four primary sources of documentation for MATPOWER.
    - MATPOWER User's Manual
    - MOST User's Manual
    - MATPOWER Online Function Reference
      (http://www.pserc.cornell.edu/matpower/docs/ref)
    - MATLAB's 'help' command

The MATPOWER and MOST User's Manuals are included in the distribution
(docs/MATPOWER-manual.pdf and docs/MOST-manual.pdf) or they can be downloaded
separately from http://www.pserc.cornell.edu/matpower/MATPOWER-manual.pdf and
http://www.pserc.cornell.edu/matpower/MOST-manual.pdf. Previous
versions are available at http://www.pserc.cornell.edu/matpower/docs/.

Each M-file has its own documentation which can be accessed by typing at
the MATLAB prompt:

    help <name of M-file>

Documentation for the case data file format can be found by typing:

    help caseformat

If something is still unclear after checking the manual and the help,
the source code *is* the documentation. ;-)

TECH NOTES

There are also two MATPOWER Technical Notes that may be of interest:

[TN1]  R. D. Zimmerman, "Uniform Price Auctions and Optimal Power Flow,
       MATPOWER Technical Note 1, February 2010.
          http://www.pserc.cornell.edu/matpower/TN1-OPF-Auctions.pdf

[TN2]  R. D. Zimmerman, "AC Power Flows, Generalized OPF Costs and
       their Derivatives using Complex Matrix Notation", MATPOWER
       Technical Note 2, February 2010.
          http://www.pserc.cornell.edu/matpower/TN2-OPF-Derivatives.pdf

PUBLICATIONS & PRESENTATIONS

[1]  R. D. Zimmerman, C. E. Murillo-Sanchez, and R. J. Thomas,
     "MATPOWER's Extensible Optimal Power Flow Architecture," Power
     and Energy Society General Meeting, 2009 IEEE, pp. 1-7,
     July 26-30 2009.
        http://www.pserc.cornell.edu/matpower/MATPOWER-OPF.pdf
        http://dx.doi.org/10.1109/PES.2009.5275967
     slides of presentation:
       http://www.pserc.cornell.edu/matpower/MATPOWER-OPF-slides.pdf

[2]  R. D. Zimmerman, C. E. Murillo-Sanchez, and R. J. Thomas,
     "MATPOWER: Steady-State Operations, Planning and Analysis Tools
     for Power Systems Research and Education," Power Systems, IEEE
     Transactions on, vol. 26, no. 1, pp. 12–19, Feb. 2011.
        http://www.pserc.cornell.edu/matpower/MATPOWER-paper.pdf
        http://dx.doi.org/10.1109/TPWRS.2010.2051168

[3]  H. Wang, C. E. Murillo-Sánchez, R. D. Zimmerman, R. J. Thomas,
     "On Computational Issues of Market-Based Optimal Power Flow,"
     Power Systems, IEEE Transactions on, vol. 22, no. 3,
     pp. 1185-1193, Aug. 2007.
        http://dx.doi.org/10.1109/TPWRS.2007.901301

[4]  C. E. Murillo-S ́anchez, R. D. Zimmerman, C. L. Anderson, and
     R. J. Thomas, "Secure Planning and Operations of Systems with
     Stochastic Sources, Energy Storage and Active Demand," Smart Grid,
     IEEE Transactions on, vol. 4, no. 4, pp. 2220–2229, Dec. 2013.
        http://dx.doi.org/10.1109/TSG.2013.2281001


-------------------
 OPTIONAL PACKAGES
-------------------

There are three optional packages to enhance the performance of MATPOWER
that may be downloaded separately. MINOPF and BPMPDMEX have more
restrictive licenses than MATPOWER. Please see the individual
Terms of Use for details.

 - BPMPD_MEX   MEX-file version of the high performance BPMPD interior
               point LP and QP solver. Speeds up DC and LP-based OPF
               solvers, and improves robustness of MINOPF.
               See http://www.pserc.cornell.edu/bpmpd/

 - CLP         COIN-OR Linear Programming solver implements high performance
               simplex and barrier LP and QP solvers that MATPOWER can use
               for the DC OPF. Available from:
               http://www.coin-or.org/projects/Clp.xml.

 - CPLEX       Includes high-performance, large-scale LP and QP solvers
               that MATPOWER can use for the DC OPF. Requires the
               Matlab interface to CPLEX, available from
               http://www.ibm.com/software/integration/optimization/cplex-optimizer/.

 - GLPK        GNU Linear Programming Kit includes large-scale LP solvers
               that MATPOWER can use for the DC OPF. Available from
               http://www.gnu.org/software/glpk/ and included with Octave.

 - GUROBI      Includes high-performance, large-scale LP and QP solvers
               that MATPOWER can use for the DC OPF. Requires the
               Gurobi MEX Matlab interface, available from
               http://www.convexoptimization.com/wikimization/index.php/Gurobi_mex.

 - IPOPT       An interior point optimizer for large scale non-linear
               optimization that MATPOWER can use for both AC and DC
               OPF problems. Requires the Matlab MEX interface for
               IPOPT, available from
               http://www.coin-or.org/projects/Ipopt.xml.
               Pre-built MEX binaries for Windows available with
               OPTI Toolbox from http://www.i2c2.aut.ac.nz/Wiki/OPTI/,
               and high-performance IPOPT-PARDISO pre-built MEX binaries
               for Mac and Linux from the PARDISO Project at
               http://www.pardiso-project.org/.

 - KNITRO      A general purpose optimization solver specializing in
               nonlinear problems that MATPOWER can use for AC OPFs.
               Requires the Knitro libraries, available from
               http://www.ziena.com/.

 - MINOPF      A MINOS-based AC OPF solver implemented as a Fortran MEX file.
               See http://www.pserc.cornell.edu/minopf/

 - MOSEK       Includes high-performance, large-scale LP and QP solvers
               that MATPOWER can use for the DC OPF. Requires the Matlab
               interface to MOSEK, available from http://www.mosek.com/.

 - TSPOPF      A package of three AC OPF solvers implemented as C MEX files.
               Suitable for large scale problems.
               See http://www.pserc.cornell.edu/tspopf/

These packages are distributed separately since each has it's own
license agreement and terms of use.

Note: For Windows users, one of the best ways to get access to some high
performance solvers, without dealing with the pain of building the MEX
interfaces yourself, is to install the OPTI Toolbox by Jonathan Currie,
available at: http://www.i2c2.aut.ac.nz/Wiki/OPTI/. The installation is
simple and it includes pre-built MEX files for several of the above
solvers, including CLP, GLPK and IPOPT.


---------------
 MAILING LISTS
---------------

There are three MATPOWER e-mail lists available to serve the MATPOWER
community:

- MATPOWER-ANNOUNCE-L is a simple announcement list for those who
  wish to be notified of the release of new versions of MATPOWER.

- MATPOWER-L is for MATPOWER users, to facilitate discussion and
  provide a forum for help with MATPOWER related questions.

- MATPOWER-DEV-L is for MATPOWER developers, to provide a forum for
  discussion related to the development of the MATPOWER software or
  proposed contributions.

For details see the Mailing Lists section of the MATPOWER website.

Please select the most appropriate list for your post and do *not*
cross-post to both MATPOWER-L and MATPOWER-DEV-L. Bug reports,
software patches, proposed enhancements, etc. should be submitted to
the issue tracker on GitHub.


CHANGES
-----------------------------
 Change history for MATPOWER
-----------------------------

Version:    6.0

maintained by Ray Zimmerman <rz10@cornell.edu>

MATPOWER
Copyright (c) 1996-2017, Power Systems Engineering Research Center (PSERC)
See http://www.pserc.cornell.edu/matpower/ for more info.

For change history for MOST, see most/docs/CHANGES.


Changes since last release
--------------------------

1/16/17
    - Support plotting of multiple nose curves in CPF by allowing
      option 'cpf.plot.bus' to take on vector values.

1/14/17
    - Add line for curtailed load to case_info() output.

1/5/17
    - Fix bug #4 where some Q limits were not being respected by CPF
      when buses were converted to PQ by initial power flow run.
      (thanks to Shruti Rao)

1/4/17
    - When genfuel field is present in mpc, load2disp() now augments
      it with 'dl' entries for the dispatchable loads it adds.

1/3/17
    - Change default implementation of active power line flow
      constraints (opf.flow_lim = 'P') to use flow directly, rather
      than square of flow, which is now a separate option, namely
      opf.flow_lim = '2'. (thanks to Nico Meyer-Huebner)

12/29/16
    - Fix bug in converting older versions of MATPOWER options struct.

12/21/16
    - Add Travis-CI integration (thanks to Richard Lincoln).


Changes since 6.0b2
-------------------

12/16/16
    - Released 6.0.
    - Moved development to GitHub <https://github.com/MATPOWER/matpower>.

12/9/16
    - Bumped version to 1.2.2.
    - Renamed MIPS from MATLAB Interior Point Solver to MATPOWER Interior
      Point Solver.

12/6/16
    - Remove dependence of t_mpsolve() on presence of have_fcn() to
      detect PARDISO installation.
    - Remove dependence of mpver() on presence of mostver().


Changes since 6.0b1
-------------------

11/1/16
    - Released 6.0b2.

10/26/16
    - Include support for MOSEK 8, some algorithms eliminated, others changed
      algorithm code set via 'mosek.lp_alg' option.

10/19/16
    - Improved robustness of infeasibility detection when 'pf.enforce_q_lims'
      option is used.
    - Add tests for power flow with 'pf.enforce_q_lims' option.

10/14/16
    - Fixed bugs in AC OPF solver code for fmincon, IPOPT, and Knitro
      involving shadow prices on variable bounds, when upper and lower bounds
      are equal (such as a voltage magnitude setpoint when 'opf.use_vg' is 1).
      In IPOPT the prices are missing, and Knitro and fmincon can both return
      negative prices on the wrong constraint.

10/13/16
    - Added 'opf.use_vg' option to provide a convenient way to have the
      OPF respect the generator voltage setpoints specified in the gen
      matrix.

10/4/16
    - Improve backward compatibility of mpoption(), allow it to handle
      old-style option vectors from earlier versions of MATPOWER, namely
      3.2 and 4.0, and to accept on old style options vector with new-style
      name/value pair option overrides.
    - INCOMPATIBLE CHANGE: Remove 'cpf.user_callback_args' option and
      modify 'cpf.user_callback' to allow for structs defining callback
      priority and args, in addition to just callback function name.
    - Add user options for setting tolerances for target lambda detection
      and nose point detection, 'cpf.target_lam_tol' and 'cpf.nose_tol',
      respectively.

9/27/16
    - INCOMPATIBLE CHANGE: Changed name of 'cpf.error_tol' option to
      'cpf.adapt_step_tol'.

9/22/16
    - Update tests for compatibility with Optimization Toolbox 7.5 (R2016b),
      which removes 'active-set' algorithm for quadprog() and 'active-set'
      and 'simplex' for linprog().
    - Fix fatal error when using some SDP_PF functions when CPLEX is not
      installed.

9/19/16
    - Fix dimension bug in makeBdc() when last bus is not connected (should
      never happen if bus is properly marked as type NONE). Thanks to
      Samuel Perkin.

8/15/16
    - Fix a harmless bug in @opt_model where variable, constraint and
      cost sets indexed by a single variable would allocate a square
      matrix of starting and ending indices, rather than a simple vector.
      Thanks to Alberto Lamadrid for catching this.

7/18/16
    - Extract computation of tangent vector from cpf_predictor() into
      cpf_tangent(), and call sequentially as needed.

7/11/16
    - Add step size to CPF callback args and results logging.
    - Add option 'cpf.adapt_step_damping' to control oscillations in
      adaptive step size control for continuation power flow.

7/8/16
    - Updated network reduction code to handle cases with radially connected
      external buses.

7/1/16
    - Fix bug in cpf_default_callback() that sometimes resulted in plotting
      voltage at different buses for different parts of the curve when the
      bus was not explicitly specified with the 'cpf.plot.bus' option.

6/23/16
    - Bad bus numbers no longer cause a fatal error (after reporting the
      bad bus numbers) in case_info().
    - Updated versions of qcqp_opf() and qcqp_opf() in extras/misc, from
      Cedric Josz.

6/10/16
    - Fix bug in savecase() where single quotes were not escaped properly
      in bus names.
    - Generator capability curve parameters that define a zero-reactive
      power line no longer cause a fatal error.


Changes since 5.1
-----------------

6/1/16
    - Released 6.0b1.

5/27/16
    - INCOMPATIBLE CHANGE: Removed fairmax() from the public interface
      by moving it inside uopf(), the only place it was used.

5/25/16
    - Add contributed code from Camille Hamon to extras/maxloadlim for
      finding the loadability limits in power systems based on an optimal
      power flow using dispatchable loads.

5/24/16
    - Fix bug in toggle_dclines() that resulted in fatal error when
      used with OPF with reactive power costs. Thanks to Irina Boiarchuk.
    - Add option to call total_load() with full case struct, instead
      of separate bus and gen matrices.

5/18/16

lib/mpver.m
function rv = mpver(varargin)
%MPVER  Prints or returns MATPOWER version info for current installation.
%   V = MPVER returns the current MATPOWER version number.
%   V = MPVER('all') returns a struct with the fields Name, Version,
%   Release and Date (all strings). Calling MPVER without assigning the
%   return value prints the version and release date of the current
%   installation of MATPOWER, MATLAB (or Octave), the Optimization Toolbox,
%   MIPS and any optional MATPOWER packages.

%   MATPOWER
%   Copyright (c) 2005-2015, Power Systems Engineering Research Center (PSERC)
%   by Ray Zimmerman, PSERC Cornell
%
%   This file is part of MATPOWER.
%   Covered by the 3-clause BSD License (see LICENSE file for details).
%   See http://www.pserc.cornell.edu/matpower/ for more info.

%% the following only works if MATPOWER is explicitly in the path,
%% but not if it is only in the current working directory
% fs = filesep;
% p = fileparts(which('runpf'));
% if ~strcmp(p(1),fs)
%   [t, p] = strtok(p, filesep);
% end
% p = p(2:end);
% v{1} = ver(p);

v{1} = struct(  'Name',     'MATPOWER', ... 
                'Version',  '6.0', ...
                'Release',  '', ...
                'Date',     '16-Dec-2016' );
if nargout > 0
    if nargin > 0
        rv = v{1};
    else
        rv = v{1}.Version;
    end
else
    if have_fcn('octave')
        v{2} = ver('octave');
    else
        v{2} = ver('matlab');
        if length(v{2}) > 1
            warning('The built-in VER command is behaving strangely, probably as a result of installing a 3rd party toolbox in a directory named ''matlab'' on your path. Check each element of the output of ver(''matlab'') to find the offending toolbox, then move the toolbox to a more appropriately named directory.');
            v{2} = v{2}(1);
        end
    end
    v{3} = ver('optim');
    if length(v{3}) > 1
        warning('The built-in VER command is behaving strangely, probably as a result of installing a 3rd party toolbox in a directory named ''optim'' on your path. Check each element of the output of ver(''optim'') to find the offending toolbox, then move the toolbox to a more appropriately named directory.');
        v{3} = v{3}(1);
    end
    for n = 1:3
        if n == 3
            if isempty(v{3}) || isempty(v{3}.Version)
                if have_fcn('matlab')
                    fprintf('\n%-22s -- not installed --', 'Optimization Toolbox');
                else    %% Octave
                    fprintf('\n%-22s -- not installed --', 'optim');
                end
                continue;
            elseif have_fcn('matlab') && ~license('test', 'optimization_toolbox')
                fprintf('\n%-22s -- no license --', 'Optimization Toolbox');
                continue;
            end
        end
        fprintf('\n%-22s Version %-9s', v{n}.Name, v{n}.Version);
        if ~isempty(v{n}.Date)
            fprintf('  %11s', v{n}.Date);
            if ~isempty(v{n}.Release)
                fprintf('   Release: %-10s', v{n}.Release);
            end
        end
    end
    fprintf('\n');
    mipsver;
    if have_fcn('most')
        mostver;
    else
        fprintf('%-22s -- not installed --\n', 'MOST');
    end
    if have_fcn('sdp_pf')
        sdp_pf_ver;
    else
        fprintf('%-22s -- not installed --\n', 'SDP_PF');
    end
    if have_fcn('yalmip')
        s = have_fcn('yalmip', 'all');
        fprintf('%-22s Version %-10s %-11s\n', 'YALMIP', s.vstr, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'YALMIP');
    end
    if have_fcn('bpmpd')
        if exist('bpver', 'file') == 2
            bpver;
        else
            fprintf('%-22s Version 2.21 or earlier\n', 'BPMPD_MEX');
        end
    else
        fprintf('%-22s -- not installed --\n', 'BPMPD_MEX');
    end
    if have_fcn('clp')
        s = have_fcn('clp', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'CLP', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'CLP');
    end
    if have_fcn('cplex')
        s = have_fcn('cplex', 'all');
        fprintf('%-22s Version %-10s %-11s\n', 'CPLEX', s.vstr, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'CPLEX');
    end
    if have_fcn('glpk')
        s = have_fcn('glpk', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'GLPK', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'GLPK');
    end
    gurobiver;
    if have_fcn('ipopt')
        s = have_fcn('ipopt', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'IPOPT', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'IPOPT');
    end
    if have_fcn('knitro')
        s = have_fcn('knitro', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'KNITRO', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'KNITRO');
    end
    if have_fcn('minopf')
        if exist('minopfver', 'file') == 2
            minopfver;
        else
            fprintf('%-22s Version 3.0b2 or earlier\n', 'MINOPF');
        end
    else
        fprintf('%-22s -- not installed --\n', 'MINOPF');
    end
    if have_fcn('mosek')
        s = have_fcn('mosek', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'MOSEK', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'MOSEK');
    end
    if have_fcn('pardiso')
        s = have_fcn('pardiso', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'PARDISO', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'PARDISO');
    end
    if have_fcn('pdipmopf')
        pdipmopfver;
    else
        fprintf('%-22s -- not installed --\n', 'PDIPMOPF');
    end
    if have_fcn('scpdipmopf')
        scpdipmopfver;
    else
        fprintf('%-22s -- not installed --\n', 'SCPDIPMOPF');
    end
    if have_fcn('sdpt3')
        s = have_fcn('sdpt3', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'SDPT3', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'SDPT3');
    end
    if have_fcn('sedumi')
        s = have_fcn('sedumi', 'all');
        if isempty(s.vstr)
            vn = '<unknown>';
        else
            vn = s.vstr;
        end
        fprintf('%-22s Version %-10s %-11s\n', 'SeDuMi', vn, s.date);
    else
        fprintf('%-22s -- not installed --\n', 'SeDuMi');
    end
    if have_fcn('tralmopf')
        tralmopfver;
    else
        fprintf('%-22s -- not installed --\n', 'TRALMOPF');
    end

    fprintf('%-22s %s\n\n', 'Architecture:', computer);
    
    fprintf('  MATPOWER %s is distributed under the 3-clause BSD License.\n', v{1}.Version);
    fprintf('  Please see the LICENSE file for details.\n\n');
end
